File: 207-course-schedule/course-schedule.py
class Solution(object):
    def canFinish(self, numCourses, prerequisites):
        preMap = {i:[] for i in range(numCourses)}
        for course, pre in prerequisites:
            preMap[course].append(pre)
        
        visitSet = set()
        def dfs(course):
            if course in visitSet:
                return False
            if preMap[course] == []:
                return True
            visitSet.add(course)
            
            for pre in preMap[course]:
                if not dfs(pre): return False
            visitSet.remove(course)
            preMap[course]=[]
            return True
        for course in range(numCourses):
            if not dfs(course): return False
        return True

        

File: 207-course-schedule/course-schedule.cpp
class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        queue<int> q;
        vector<vector<int>> adj(numCourses);
        vector<int> indegree(numCourses, 0);

        for(auto &n: prerequisites){
            adj[n[1]].push_back(n[0]);
            indegree[n[0]]++;
        }
        
        for(int i=0; i<numCourses; i++){
            if(indegree[i]==0){
                q.push(i);
            }
        }
        int count=0;
        while(!q.empty()){
            int top=q.front();
            q.pop();
            for(int ad: adj[top]){
                indegree[ad]--;
                if(indegree[ad]==0){
                    q.push(ad);
                }
            }
            count++;
            
        }
    return count==numCourses;
    }
};